src/net/sourceforge/kolmafia/request/ClanLoungeRequest.java

```diff
@@ -340,94 +324,202 @@ public static final boolean isFancyHotDog(String name) {
+   *    concoction.resetCalculations() for all speakeasy drinks.
+   *
+   * *** Current implementation adds or removes speakeasy drinks to the usables list.
+   * *** That is incorrect; the list should remain untouched and visibility of the drink
+   * *** determined by the ListCellRenderer
+   */
+
+  private static final Set<SpeakeasyDrink> allSpeakeasyDrinks = new HashSet<>();
+
+  public static class SpeakeasyDrink {
+    private static final Map<String, SpeakeasyDrink> nameToData = new HashMap<>();
+    private static final Map<String, String> canonicalNameToName = new HashMap<>();
+    private static final Map<Integer, SpeakeasyDrink> idToData = new HashMap<>();
+
+    public static SpeakeasyDrink findName(String name) {
+      return SpeakeasyDrink.nameToData.get(name);
     }
-    return -1;
-  }
 
-  public static final String speakeasyIdToName(int id) {
-    int index = ClanLoungeRequest.speakeasyIdToIndex(id);
-    return index < 0 ? null : ClanLoungeRequest.SPEAKEASY_DATA[index].name;
-  }
+    public static String findCanonicalName(String canonicalName) {
+      return SpeakeasyDrink.canonicalNameToName.get(canonicalName);
+    }
 
-  public static final String speakeasyIndexToName(int index) {
-    return (index < 0 || index > ClanLoungeRequest.SPEAKEASY_DATA.length)
-        ? null
-        : ClanLoungeRequest.SPEAKEASY_DATA[index].name;
-  }
+    public static SpeakeasyDrink findId(int id) {
+      return SpeakeasyDrink.idToData.get(id);
+    }
 
-  public static final Integer speakeasyIndexToId(int index) {
-    return (index < 0 || index > ClanLoungeRequest.SPEAKEASY_DATA.length)
-        ? -1
-        : ClanLoungeRequest.SPEAKEASY_DATA[index].id;
-  }
+    // Class fields
+    private final String name;
+    private final int id;
+    private final int inebriety;
+    private final int cost;
 
-  public static final Integer speakeasyIndexToInebriety(int index) {
-    return (index < 0 || index > ClanLoungeRequest.SPEAKEASY_DATA.length)
-        ? -1
-        : ClanLoungeRequest.SPEAKEASY_DATA[index].inebriety;
+    // Derived fields
+    private final AdventureResult item;
+    private final String canonicalName;
+    private final Concoction concoction;
+
+    public SpeakeasyDrink(String name, int id, int inebriety, int cost) {
+      this.name = name;
+      this.id = id;
+      this.inebriety = inebriety;
+      this.cost = cost;
+
+      // Derived fields
+      this.canonicalName = StringUtilities.getCanonicalName(name);
+
+      this.item = ItemPool.get(ItemDatabase.getItemId(name, 1, false));
+
+      Concoction concoction = ConcoctionPool.get(this.item);
+      concoction.speakeasy = true;
+      concoction.price = cost;
+      this.concoction = concoction;
+
+      // Lookups
+      SpeakeasyDrink.nameToData.put(name, this);
+      SpeakeasyDrink.canonicalNameToName.put(canonicalName, name);
+      SpeakeasyDrink.idToData.put(id, this);
+
+      ClanLoungeRequest.allSpeakeasyDrinks.add(this);
```

Comment:
while this is a pattern that occurs elsewhere in the Mafia codebase, I really would much prefer it if the constructor didn't have any side effects and the static initialization block at line 463, instead of just calling each constructor, called something like `populateSpeakeasyData()` with the relevant values (for each drink) which did the setup.

Reply:
This was previously a record. I rejected that since those are not allow to have private - "derived" - fields.
Both a class and an enum allow those.
Additionally, both allow private static members.
gausie had a class (or enum?) with a private static map and a public static "findInMap" method.

I quite liked that, since the maps were private to the inner class, rather than visible (public or private) in the enclosing class, and anybody with access to class could call the static class method to lookup an object.

Do you object to private static class members in inner classes?
If not, surely they have to be manipulated in the constructor?
It is a "side effect".

You can do the same thing with enums - and, in a case like this, where all the members are known at compile time, that is probably a better choice. The global map could be a EnumSet.allOf(SpeakeasyDrink).

I will try converting to an Enum and we can see what it looks like then.

If the issue is that you don't think inner classes should have static members, we will probably have to agree to disagree. :)

Reply:
> If not, surely they have to be manipulated in the constructor?

or from methods called off the object, outside the constructor?

> It is a "side effect".

I suppose the main reason I don't like this style of construction is that it has the following code:

```
static {
  new SpeakeasyDrink("glass of &quot;milk&quot;", 1, 1, 250);
  new SpeakeasyDrink("cup of &quot;tea&quot;", 2, 1, 250);
  new SpeakeasyDrink("thermos of &quot;whiskey&quot;", 3, 1, 250);
  new SpeakeasyDrink("Lucky Lindy", 4, 1, 500);
  new SpeakeasyDrink("Bee's Knees", 5, 2, 500);
  new SpeakeasyDrink("Sockdollager", 6, 2, 500);
  new SpeakeasyDrink("Ish Kabibble", 7, 2, 500);
  new SpeakeasyDrink("Hot Socks", 8, 3, 5000);
  new SpeakeasyDrink("Phonus Balonus", 9, 3, 10000);
  new SpeakeasyDrink("Flivver", 10, 2, 20000);
  new SpeakeasyDrink("Sloppy Jalopy", 11, 5, 100000);
}
```

We create a series of `SpeakeasyDrink`s, and then throw them all away. Without knowing that the constructor adds all constructed objects to a list, this seems quite surprising.

Additionally, it means that you can't create a new `SpeakeasyDrink` object, because the constructor does something in addition to merely constructing the object.

There's [some discussion on StackOverflow](https://softwareengineering.stackexchange.com/questions/343447/should-constructors-ever-be-used-only-for-side-effects) about using constructors for side effects: the answer references SRP and the principle of least astonishment. I was never too big on SRP as I saw it massively overused to fragment classes into far-too-tiny portions, but I think the principle of least astonishment has some merit. I expect a constructor to construct an object, and find it surprising if it also does something else (e.g. adding the object to a list).

Reply:
OK. SpeakeasyDrink is now an enum with no side effects in its constructor.
See how you like it.